src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -0,0 +1,350 @@
+        {
+            return CollectDynamicAnalysis(base.VisitStatementList(node));
+        }
+
+        public override BoundNode VisitSwitchStatement(BoundSwitchStatement node)
+        {
+            return CollectDynamicAnalysis(base.VisitSwitchStatement(node));
+        }
+
+        public override BoundNode VisitThrowStatement(BoundThrowStatement node)
+        {
+            return CollectDynamicAnalysis(base.VisitThrowStatement(node));
+        }
+
+        public override BoundNode VisitTryStatement(BoundTryStatement node)
+        {
+            return CollectDynamicAnalysis(base.VisitTryStatement(node));
+        }
+
+        public override BoundNode VisitUsingStatement(BoundUsingStatement node)
+        {
+            return CollectDynamicAnalysis(base.VisitUsingStatement(node));
+        }
+
+        public override BoundNode VisitWhileStatement(BoundWhileStatement node)
+        {
+            return CollectDynamicAnalysis(base.VisitWhileStatement(node));
+        }
+
+        public override BoundNode VisitYieldBreakStatement(BoundYieldBreakStatement node)
+        {
+            return CollectDynamicAnalysis(base.VisitYieldBreakStatement(node));
+        }
+
+        public override BoundNode VisitYieldReturnStatement(BoundYieldReturnStatement node)
+        {
+            return CollectDynamicAnalysis(base.VisitYieldReturnStatement(node));
+        }
+
+        private BoundNode CollectDynamicAnalysis(BoundNode node)
+        {
+            BoundStatement statement = node as BoundStatement;
+            if (statement != null)
+            {
+                return CollectDynamicAnalysis(statement);
+            }
+
+            return node;
+        }
+
+        private BoundNode CollectDynamicAnalysis(BoundStatement statement)
+        {
+            if (statement.WasCompilerGenerated)
+            {
+                return statement;
+            }
+
+            Location statementLocation = statement.Syntax.GetLocation();
+            FileLinePositionSpan lineSpan = statementLocation.GetMappedLineSpan();
+            string path = lineSpan.Path;
+            if (path == "")
+            {
+                path = statement.Syntax.SyntaxTree.FilePath;
+            }
+
+            int spansIndex = _spansBuilder.Count;
+            _spansBuilder.Add(new SourceSpan(_debugDocumentProvider.Invoke(path, ""), lineSpan.StartLinePosition.Line, lineSpan.StartLinePosition.Character, lineSpan.EndLinePosition.Line, lineSpan.EndLinePosition.Character));
+
+            // Generate "_payload[pointIndex] = true".
+
+            SyntheticBoundNodeFactory statementFactory = new SyntheticBoundNodeFactory(_method, statement.Syntax, _compilationState, _diagnostics);
+            BoundArrayAccess payloadCell = statementFactory.ArrayAccess(statementFactory.Field(null, _payload), statementFactory.Literal(spansIndex));
+            BoundExpressionStatement cellAssignment = statementFactory.Assignment(payloadCell, statementFactory.Literal(true));
```

Comment:
Still thinking about this. The details of instrumentation will change in various ways, and I'll work through it when there is a little more time to ponder things.